pages/api/api-mocker/[...slug].ts

```diff
@@ -0,0 +1,199 @@
+
+  const regex = new RegExp("^" + regexPath + "$");
+
+  const res = regex.test(slug);
+  return res;
+}
+
+export function processMultipleMatches(
+  matches: Action[],
+  slug: string[]
+): Action[] {
+  // slug = 'api/v1/Customers/location'. matches = /api/v1/Customers/location and /api/v1/Customers/{id}.
+  const matchEnd = matches.filter((match) => {
+    const split = match.path?.split("/").filter((arr) => arr != "") ?? [];
+    return split[split.length - 1] === slug[slug.length - 1];
+  });
+  if (matchEnd.length === 1) return matchEnd;
+
+  // slug = 'api/v2/Coordinators/1234'. matches = "/api/v2/Coordinators/{id}". "/api/v2/Coordinators/location"
+  const matchPenultimate = matches.filter((match) => {
+    const split = match.path?.split("/").filter((arr) => arr != "") ?? [];
+    return (
+      split[split.length - 1].includes("}") &&
+      split[split.length - 1].includes("{") &&
+      split[split.length - 2] === slug[slug.length - 2]
+    );
+  });
+  if (matchPenultimate.length === 1) return matchPenultimate;
+
+  throw new Error(
+    `Failed to match slug: "${slug}" to single action. Candidate paths: ${matches.map(
+      (match) => match.path
+    )}`
+  );
+}
+
+export function getMatchingAction(
+  org_id: number,
+  actions: Action[],
+  method: RequestMethods,
+  slug: string[]
+): Action | null {
+  const slugString = "/" + slug.join("/");
+
+  let matches = actions.filter((action) => {
+    return action.path && slugMatchesPath(action.path, slugString);
+  });
+
+  if (matches.length === 0) {
+    console.log(
+      `Could not match slug "${slugString}" with method "${method}" to any action in the database for organisation "${org_id}"`
+    );
+    return null;
+  }
+
+  if (matches.length > 1) matches = processMultipleMatches(matches, slug);
+
+  console.log(
+    `Successfully matched slug "${slugString}" to action with path "${matches[0].path}"`
+  );
+
+  return matches[0];
+}
+
+export function getPathParameters(
+  path: string,
+  querySlugs: string[]
+): { [name: string]: string } {
+  let variables: { [name: string]: string } = {};
+  const urlSlugs = path.split("/").filter((slug) => slug !== "");
+  if (urlSlugs.length !== querySlugs.length)
+    throw new Error(
+      `Path and query slugs lengths do not match: [${urlSlugs}] AND  [${querySlugs}]`
+    );
+
+  for (let index = 0; index < urlSlugs.length; index++) {
+    // slicing to remove { and } from urlSlugs variable name
+    if (urlSlugs[index].startsWith("{") && urlSlugs[index].endsWith("}")) {
+      const variableName = urlSlugs[index].slice(1, urlSlugs[index].length - 1);
+      variables[variableName] = querySlugs[index];
+    }
+  }
+  return variables;
+}
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<void> {
+  const queryParams = req.query;
+  const org_id = Number(req.headers["org_id"]);
+  const slug = queryParams.slug as string[];
```

Comment:
We need some form of authentication on this endpoint - we don't want it accessible to the internet :smile: